presto-main/src/test/java/io/prestosql/sql/query/TestUnwrapCastInComparison.java

```diff
@@ -0,0 +1,243 @@
+        List<String> values = ImmutableList.of("-999999999999999", "999999999999999");
+        for (String from : values) {
+            for (String operator : Arrays.asList("=", "<>", ">=", ">", "<=", "<", "IS DISTINCT FROM")) {
+                for (String to : values) {
+                    assertTrue(evaluate(operator, "DECIMAL(15, 0)", from, "DOUBLE", Double.valueOf(to)));
+                }
+            }
+        }
+
+        // decimal(16) -> double
+        values = ImmutableList.of("-9999999999999999", "9999999999999999");
+        for (String from : values) {
+            for (String operator : Arrays.asList("=", "<>", ">=", ">", "<=", "<", "IS DISTINCT FROM")) {
+                for (String to : values) {
+                    assertTrue(evaluate(operator, "DECIMAL(16, 0)", from, "DOUBLE", Double.valueOf(to)));
+                }
+            }
+        }
+
+        // decimal(7) -> real
+        values = ImmutableList.of("-9999999", "9999999");
+        for (String from : values) {
+            for (String operator : Arrays.asList("=", "<>", ">=", ">", "<=", "<", "IS DISTINCT FROM")) {
+                for (String to : values) {
+                    assertTrue(evaluate(operator, "DECIMAL(7, 0)", from, "REAL", Double.valueOf(to)));
+                }
+            }
+        }
+
+        // decimal(8) -> real
+        values = ImmutableList.of("-99999999", "99999999");
+        for (String from : values) {
+            for (String operator : Arrays.asList("=", "<>", ">=", ">", "<=", "<", "IS DISTINCT FROM")) {
+                for (String to : values) {
+                    assertTrue(evaluate(operator, "DECIMAL(8, 0)", from, "REAL", Double.valueOf(to)));
+                }
+            }
+        }
+    }
+
+    @Test
+    public void testVarchar()
+    {
+        for (String from : Arrays.asList("", "a", "b")) {
+            for (String operator : Arrays.asList("=", "<>", ">=", ">", "<=", "<", "IS DISTINCT FROM")) {
+                for (String to : Arrays.asList("", "a", "aa", "b", "bb")) {
+                    assertTrue(evaluate(operator, "VARCHAR(1)", from, "VARCHAR(2)", to));
+                }
+            }
+        }
+
+        // type with no range
+        for (String operator : Arrays.asList("=", "<>", ">=", ">", "<=", "<", "IS DISTINCT FROM")) {
+            for (String to : Arrays.asList(Strings.repeat("a", 200), Strings.repeat("b", 200))) {
+                assertTrue(evaluate(operator, "VARCHAR(200)", Strings.repeat("a", 200), "VARCHAR(300)", to));
+            }
+        }
+    }
+
+    public boolean evaluate(String operator, String fromType, Object fromValue, String toType, Object toValue)
+    {
+        // Use if(rand() >= 0, ...) to defeat the constant folder
+        String query = format(
+                "VALUES (CAST(IF(rand() >= 0, CAST('%s' AS %s)) AS %s) %s CAST('%s' AS %s)) " +
+                        "IS NOT DISTINCT FROM " +
+                        "(CAST('%s' AS %s) %s CAST('%s' AS %s))",
```

Comment:
Just a thought. There is a way to avoid using `rand` hack, but it wouldn't be as terse.
You could use `ExpressionInterpreter` to evaluate cast independently and then literalEncode + construct Comparison and evaluate that.
(no change requested)